Doc/regions.md

```diff
@@ -0,0 +1,567 @@
+Here we present the idea as a reverse polish evaluator.
+```python
+>>> x = 1
+>>> y = 2
+>>> calc("4 x y + *")
+```
+The idea is to interpolate the `x` and `y` into the string while evaluating it.
+```
+def calc(str):
+  ctxt = sys._getframe().f_back
+  ops = str.split()
+  stack = []
+  for op in ops:
+    try: 
+      stack.append(int(op))
+    except ValueError:
+      if op == "+":
+        stack.append(stack.pop() + stack.pop())
+      elif op == "*":
+        stack.append(stack.pop() * stack.pop())
+      else:    
+        stack.append(ctxt.f_locals[op])      
+  return stack.pop()
+```
+This is a simple example, but it shows how the frame object can be used to access the local variables in the calling scope.
+```
+>>> x = 1
+>>> y = 2
+>>> calc("4 x y + *")
+12
+```
+
+The numpy library is pretty popular, so this is a good example of how to use the frame object in a way that is not debugging.
+
+
+Note that the frame object is mostly read-only, but does contain an interesting operation to set the executing line number of a frame.
+```C
+static int
+frame_setlineno(PyFrameObject *f, PyObject* p_new_lineno, void *Py_UNUSED(ignored))
+```
+This seems very bad to call from another thread/interpreter.
+
+Ultimately, a frame object should never escape the thread/interpreter that it was created in.
+
+
+
+## Regions in Veronapy.
+
+The previous section illustrates the many temporary objects that Python creates to handle variables.
+To support this, we need our region object model to support complex local structures for indexing into regions.
+To achieve this, we will use three kinds of object ownership:
+
+* Local object 
+* Immutable
+* Regions
+
+The "local objects" are used to represent the many local structures that Python creates such as function objects, frame objects and cells.
+These objects are typically not shared between threads, so normally should not require any region tracking.
+However, we need to support local objects becoming immutable to allow them to be shared, and suppport for groups of objects to be moved between interpreters.
+
+To pass mutable states between threads, we need to introduce a new construct to Python: "regions".
+A region is a group of objects that can move together across threads
+and can be accessed by at most one interpreter at a time.
+This removes the need for locks to protect the objects in the region.
+The region can be passed between threads, and the receiving interpreter can access the objects in the region.
+
+To efficiently support regions, we impose the following invariants that the runtime is responsible for maintaining:
+
+* Local objects can reach anything
+* Immutable objects can only reach other immutable objects.
+* Objects in a region can only reach other objects in the same region, immutable objects, or linear region entry points.
```

Comment:
What's a linear region entry point? I don't see that term explained elsewhere (in this doc).

Reply:
Good catch! A region’s entry point is an object of (let’s say) the `Region` class that is the handle to a region. So if I want to pass a region somewhere, I pass that object. So creating a new region will be essentially `r = Region()` and putting something in that region will be f.ex. `r.f = x` and subsequently also `x.f = y` because `x` now points to something in the region. 

This entry point is linear means that we want to assign it / read it from fields by moving the reference, not copying the reference. 

That means that, considering heap references only, all regions form a forest of trees. That is great for reasoning about sharing and transfer of objects.

Reply:
> This entry point is linear means that we want to assign it / read it from fields by moving the reference, not copying the reference.

My head just exploded. What does this mean? In particular I am confused by "assign / read" and "moving the reference". Is that C++ move semantics? How should I understand that in a Python context?

> That means that, considering heap references only, all regions form a forest of trees. That is great for reasoning about sharing and transfer of objects.

This isn't quite as head-exploding but I am still confused. Is each region a tree? Or a forest? Or is the set of all regions, when only considering references from one region to another (understanding that such references aways point to entry points), a forest of trees?

And why is that helpful for reasoning about sharing?

I wish we were standing in front of a physical whiteboard now. :-)

Reply:
Sorry for vanishing for a bit. Grant deadline… Let's pick this up next week! 

> My head just exploded. What does this mean? In particular I am confused by "assign / read" and "moving the reference". 

Sorry! 

> Is that C++ move semantics? 

More or less yes, but see below.

> How should I understand that in a Python context?

This is a design point that is still being discussed. The strictest possible sense to interpret this is essentially that `x = y` copies the contents of `y` into `x` and then "invalidates `y`". Invalidating `y` could be undefining it, storing `None` into it, or something else. A much more relaxed position is that `x = y` works as normal, but if we ever need for the contents of `x` to be "linear" (in the sense of being the only reference to the thing it references*), then `y` needs to have been invalidated or overwritten before that point, or we need to throw an exception. 

> Is each region a tree? Or a forest? 

Think of a region as a boundary that you draw around all the objects in each region. In a good program, every region should have at most one incoming pointer from the heap (excluding pointers internal to the region itself). Thus, if we squint a bit and forget about the objects inside the regions and just look at the regions themselves, we have a forest of region trees. I hope this makes sense! Happy to make another attempt at explaining this. 

> And why is that helpful for reasoning about sharing?

If we know we are operating on different trees, we know that the operations cannot interfere with each other. 

If a structure has a tree shape, then `x.f` and `x.g` cannot alias (further, `x.f.h` and `x.g.j` also won't alias, etc.). That means that if we move the contents of `x.f into `y` (e.g. `y = x.f; x.f = None`), we can rely on the fact there is no overlap between `y` and `x`meaning that the references can be passed to different threads without risking interference. 

I hope this makes some kind of sense!

Reply:
Thanks for the response! I think I have a better understanding of the whole thing now, thanks to your response and to reading the whole doc more carefully (and also finally reading the internal Word doc that was used for the March project review).

> This is a design point that is still being discussed. The strictest possible sense to interpret this is essentially that `x = y` copies the contents of `y` into `x` and then "invalidates `y`". Invalidating `y` could be undefining it, storing `None` into it, or something else. A much more relaxed position is that `x = y` works as normal, but if we ever need for the contents of `x` to be "linear" (in the sense of being the only reference to the thing it references*), then `y` needs to have been invalidated or overwritten before that point, or we need to throw an exception.

For the strict interpretation, you would be working against all forces in Python. Python does not allow overloading assignment; plain assignment `x = <expr>` is *always* a pointer copy plus an incref, and there is never any requirement that `<expr>` must be a variable. It's more subtle for `o.x = <expr>`, where the object `o` gains control over the assignment via the `__setattr__` overload, but unless we're talking about unboxing numbers, there's at least a strong expectation that this just copies the pointer.

Invalidating the source of the assignment is even less "taboo" -- the language has no way to even talk about that kind of thing. (OTOH, Mojo does -- it introduces new syntax for parameter transfer modes, but in the end I think of Mojo as a Rust-like statically typed and compiled language with a Python-flavored syntax, not as a Python variant.)

IOW, I think that the relaxed interpretation is probably the way to go.

I do finally understand the "forest of trees" phrase -- there is only one reference to each region and there are no cycles in the graph of regions. Preserving this invariant in CPython will be hard though -- e.g. if we have some code that uses a region (e.g. a `when` block) and we want to refactor it so that some of the code is moved into a function, the obvious Pythonic pattern would be to write the function as something that takes the region as an argument, but when we call that function, we incref the region object (as we do all arguments on a call).

Example: Python users would expect these two to be equivalent:
```
when r:
    r.foo = 1
```
and
```
def func(r):
    r.foo = 1
when r:
    func(r)
```

I'm not sure what to suggest here. Maybe it's sufficient to insist on a single reference to a region at the point when we transfer it to another thread/subinterpreter/core. The receiving interpreter then owns that region and the way interpreters are isolated guarantees that no other interpreter can access the region.

Reply:
> For the strict interpretation, you would be working against all forces in Python. 
> IOW, I think that the relaxed interpretation is probably the way to go.

Yeah, that was my impression too. I mostly wrote it down for completeness and because it thought it was easier to explain the big idea that way. I believe that we can make do without the strict implementation, at least for stack variables which I think is the problem. So for simplicity (I'll complicate this further down), we can say that an object in a region can have:

1. An unlimited number of references to it from variables in the stack
2. An unlimited number of references to it from other objects in the same region
3. At most one reference to it from an object in another region

Furthermore, with respect to (3), at most one object in a region can have references to it from objects in another region. This is the object we call "the region object" and (at least for now) we require this object to be an instance of a special class (e.g. `Region`). 

With the above, a programmer could implement a move from `x.f` to `y.f` for example by doing `z = x.f; x.f = None; y.f = z`. I guess we could also add some convenience functions like `move(x, "f")` that was (morally) implemented `z = x.f; x.f = None; return z` (There most likely exists a more elegant way.). Calls to these would reduce clutter and serve as useful documentation in the code. 

As we permit both (1) and (2) in the list above, we probably don't want to have have `__getattr__` implicitly perform a move when it detects that the content of a field is a region object.  

> Example: Python users would expect these two to be equivalent:

Hmm. I think this should hold in our model too.  Maybe I am not following where the cycle is created!

In the second case, the `when` block gets hold of `r` and it should be fine to simply pass `r` as an argument (copy + incref), dereference it and write to one of its fields. In this case, the object we write to the field is immutable, which is a simple case for us. But even if it was a newly created mutable object (e.g. `x = [1]; r.foo = x`), this should work. In this case, that object would be sucked into the region. The `x` variable from the stack into the region is permitted under the relaxed rules above. Then when `func()` returns, `x` is invalidated (and `r` is decref'd), so there are no references into the region from outside of it, except `r`. Can you help me explain where in your mind the problem is, to help me understand? 

> I'm not sure what to suggest here. Maybe it's sufficient to insist on a single reference to a region at the point when we transfer it to another thread/subinterpreter/core. 

This is exactly what we are thinking!

Reply:
> An unlimited number of references to it from variables in the stack

These references could also be contained in other objects that are referenced from the stack, e.g. it's pretty common in Python to put a couple of items in a tuple (or a list, or a dict, or a user-defined object, or something more complex built by nesting some of these) and pass it around. At this point I would find it hard to tell the difference between what's referenced from the stack and what's referenced from the "locals" (in your terminology, which IIRC refers to all objects owned by a particular interpreter).

> > Example: Python users would expect these two to be equivalent:
> 
> Hmm. I think this should hold in our model too. Maybe I am not following where the cycle is created!

Sorry, I wasn't implying there was a cycle. I think I was attacking the strawman "there should be only reference to a region" which you've addressed much better with the three bullets above.

At this point I think we are completely on the same page about this!

I foresee trouble when we get to the heap-per-interpreter approach that CPython 3.12 currently uses. I suppose that's where the rubber will meet the road, since ultimately your solution there is to have an allocation strategy per region, which to me feels close to having a heap per region, which would require us to be able to move objects between heaps (but only in a lock-free, thread-safe manner, since the move would always be between a region's heap and an interpreter's "local" heap, when that interpreter has access to that region).

Reply:
> These references could also be contained in other objects that are referenced from the stack, e.g.

Yes, indeed. I didn't explain that well. We do allow this and they are permitted in my rules because those objects don't belong to a region. So morally, they are on the stack. Called "local objects" in the doc. Sorry, I didn't mean to introduce new terminology.

Reply:
> At this point I think we are completely on the same page about this!

Ah, that's great! I'm super pleased.

Reply:
Yeah, this is a good area to deviate.

> In theory this could actually improve GC by making it a lot more incremental and without global pauses.

Just know that in 3.13 we already have more incremental GC, and also, for CPython GC, theory and practice are often rather far apart...